contracts/contango/ContangoCauldron.sol

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.6;
+
+import "@yield-protocol/utils-v2/contracts/math/WMul.sol";
+import "@yield-protocol/utils-v2/contracts/math/WDivUp.sol";
+import "../constants/Constants.sol";
+import "../Cauldron.sol";
+import "./util/EnumerableSet.sol";
+
+contract ContangoCauldron is Cauldron {
+    using CauldronMath for uint128;
+    using EnumerableSet for EnumerableSet.Bytes6Set;
+    using CastU256I256 for uint256;
+    using WMul for uint256;
+    using WDivUp for uint256;
+
+    EnumerableSet.Bytes6Set private assetsInUse;
+    mapping(bytes6 => DataTypes.Balances) public balancesPerAsset;
+    int256 public peekFreeCollateralUSD;
+
+    uint128 public collateralisationRatio; // Must be on commonCcy precision
+    bytes6 public commonCcy; // Currency to use as common ground for all the ink & art
+
+    constructor(uint128 _collateralisationRatio, bytes6 _commonCurrency) {
+        collateralisationRatio = _collateralisationRatio;
+        commonCcy = _commonCurrency;
+    }
+
+    function setCollateralisationRatio(uint128 _collateralisationRatio) external auth {
+        collateralisationRatio = _collateralisationRatio;
+    }
+
+    function setCommonCurrency(bytes6 _commonCurrency) external auth {
+        commonCcy = _commonCurrency;
+    }
+
+    function pour(
+        bytes12 vaultId,
+        int128 ink,
+        int128 art
+    ) external override auth returns (DataTypes.Balances memory) {
+        (
+            DataTypes.Vault memory vault_,
+            DataTypes.Series memory series_,
+            DataTypes.Balances memory balances_
+        ) = vaultData(vaultId, true);
+
+        balances_ = _pour(vaultId, vault_, balances_, series_, ink, art);
+
+        _updateVaultBalancesPerAsset(series_, vault_, ink, art);
+
+        return balances_;
+    }
+
+    function _updateVaultBalancesPerAsset(
+        DataTypes.Series memory series,
+        DataTypes.Vault memory vault,
+        int128 ink,
+        int128 art
+    ) internal {
+        if (ink != 0) {
+            assetsInUse.add(vault.ilkId);
+            balancesPerAsset[vault.ilkId].ink = balancesPerAsset[vault.ilkId].ink.add(ink);
+        }
+        if (art != 0) {
+            assetsInUse.add(series.baseId);
+            balancesPerAsset[series.baseId].art = balancesPerAsset[series.baseId].art.add(art);
+        }
+
+        require(getFreeCollateralUSD() >= 0, "Undercollateralised");
+    }
+
+    function getFreeCollateralUSD() public returns (int256) {
+        (uint128 _collateralisationRatio, bytes6 _commonCurrency) = (collateralisationRatio, commonCcy);
+
+        uint256 totalInk;
+        uint256 totalArt;
+
+        //TODO is this reading .length() from storage each time?
```

Comment:
As far as I understand, it is, @devtooligan?

Reply:
yes should prob cash the length to save gas.  

while we're at i think we can just cash the whole `assetsInUse` to save gas on read in line 81?

and can also do something about index++ --> ideally make the increment unchecked or at least use ++index instead

Reply:
ok, I'll do a pass on gas optimisations after I have the unit tests

Reply:
Can't cache the whole `assetsInUse` as it contains nested mappings, but I addressed the increment and cached the length